src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java

```diff
@@ -29,312 +29,496 @@
+    private final ThreadFactory threadFactory;
     private final ThreadFlock flock;
-    private final ReentrantLock shutdownLock = new ReentrantLock();
 
-    // states: OPEN -> SHUTDOWN -> CLOSED
-    private static final int OPEN     = 0;   // initial state
-    private static final int SHUTDOWN = 1;
-    private static final int CLOSED   = 2;
+    // state, only accessed by owner thread
+    private static final int ST_NEW            = 0;
+    private static final int ST_FORKED         = 1;   // subtasks forked, need to join
+    private static final int ST_JOIN_STARTED   = 2;   // join started, can no longer fork
+    private static final int ST_JOIN_COMPLETED = 3;   // join completed
+    private static final int ST_CLOSED         = 4;   // closed
+    private int state;
+
+    // timer task, only accessed by owner thread
+    private Future<?> timerTask;
+
+    // set or read by any thread
+    private volatile boolean cancelled;
+
+    // set by the timer thread, read by the owner thread
+    private volatile boolean timeoutExpired;
+
+    /**
+     * Throws WrongThreadException if the current thread is not the owner thread.
+     */
+    private void ensureOwner() {
+        if (Thread.currentThread() != flock.owner()) {
+            throw new WrongThreadException("Current thread not owner");
+        }
+    }
+
+    /**
+     * Throws IllegalStateException if already joined or scope is closed.
+     */
+    private void ensureNotJoined() {
+        assert Thread.currentThread() == flock.owner();
+        if (state > ST_FORKED) {
+            throw new IllegalStateException("Already joined or scope is closed");
+        }
+    }
+
+    /**
+     * Throws IllegalStateException if invoked by the owner thread and the owner thread
+     * has not joined.
+     */
+    private void ensureJoinedIfOwner() {
+        if (Thread.currentThread() == flock.owner() && state <= ST_JOIN_STARTED) {
+            throw new IllegalStateException("join not called");
+        }
+    }
+
+    /**
+     * Interrupts all threads in this scope, except the current thread.
+     */
+    private void implInterruptAll() {
+        flock.threads()
+                .filter(t -> t != Thread.currentThread())
+                .forEach(t -> {
+                    try {
+                        t.interrupt();
+                    } catch (Throwable ignore) { }
+                });
+    }
+
+    @SuppressWarnings("removal")
+    private void interruptAll() {
+        if (System.getSecurityManager() == null) {
+            implInterruptAll();
+        } else {
+            PrivilegedAction<Void> pa = () -> {
+                implInterruptAll();
+                return null;
+            };
+            AccessController.doPrivileged(pa);
+        }
+    }
+
+    /**
+     * Cancel the scope if not already cancelled.
+     */
+    private void cancel() {
+        if (!cancelled && CANCELLED.compareAndSet(this, false, true)) {
```

Comment:
This could be simplified to:

```suggestion
        if(!CANCELLED.getAndSet(this, true))
        if (!cancelled && CANCELLED.compareAndSet(this, false, true)) {
```

Reply:
It's intentional, to avoid the CAS if possible.